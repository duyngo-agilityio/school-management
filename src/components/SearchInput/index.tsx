'use client';

import { ChangeEvent } from 'react';

// Icons
import { SearchIcon } from '@/icons';
import { TiDeleteOutline } from 'react-icons/ti';

// Components
import {
  Input,
  InputGroup,
  InputLeftElement,
  InputRightElement,
} from '@chakra-ui/react';

interface SearchInputProps {
  placeholder: string;
  defaultValue: string;
  onSearch: (value: string) => void;
  onClick: (value: string) => void;
}

const SearchInput = ({
  placeholder,
  defaultValue,
  onSearch,
  onClick,
}: SearchInputProps) => {
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    onSearch(e.target.value);
  };

  const handleClick = () => {
    onClick('');
  };

  return (
    <InputGroup width="auto">
      <InputLeftElement style={{ cursor: 'pointer' }}>
        <SearchIcon />
      </InputLeftElement>
      <Input
        bgColor="backgroundWhiteHint"
        pl={10}
        placeholder={placeholder}
        defaultValue={defaultValue}
        onChange={handleChange}
      />
      <InputRightElement>
        <TiDeleteOutline style={{ cursor: 'pointer' }} onClick={handleClick} />
      </InputRightElement>
    </InputGroup>
  );
};

export default SearchInput;
