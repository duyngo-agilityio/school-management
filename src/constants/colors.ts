const bluePallet = {
  zodiac: '#152259',
  koamaru: '#13296A',
  havelock: '#1C5A8C',
  cornflower: '#3182ce',
  solitude: '#EBF6FF',
  catSkill: '#E2E8F0',
  linkWater: '#EFF3FA',
};

const grayPallet = {
  cod: '#1A1A1A',
  tundora: '#424242',
  emperor: '#4F4F4F',
  ash: '#8A8A8A',
  slate: '#595959',
  aluminum: '#BDBDBD',
};

const whitePallet = {
  seashell: '#F1F1F1',
  snow: '#FCFAFA',
  pure: '#FFFFFF',
};

const redPallet = {
  roman: '#DB5050',
};

const yellowPallet = {
  anzac: '#DBBD50',
};

export const customColors = {
  ...bluePallet,
  ...whitePallet,
  ...grayPallet,
  ...redPallet,
  ...yellowPallet,
} as const;
