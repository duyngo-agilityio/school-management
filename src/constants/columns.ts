const COLUMNS = {
  FIELDS: {
    NAME: 'fullName',
    ID: 'id',
    EMAIL: 'email',
    CLASS: 'class',
    GENDER: 'gender',
    SUBJECT: 'subject',
  },
  HEADER_NAME: {
    NAME: 'Name',
    SUBJECT: 'Subject',
    CLASS: 'Class',
    EMAIL: 'Email Address',
    GENDER: 'Gender',
    STUDENT_ID: 'Student ID',
  },
};

export const COLUMNS_STUDENT = [
  {
    field: COLUMNS.FIELDS.NAME,
    headerName: COLUMNS.HEADER_NAME.NAME,
  },
  {
    field: COLUMNS.FIELDS.ID,
    headerName: COLUMNS.HEADER_NAME.STUDENT_ID,
  },
  {
    field: COLUMNS.FIELDS.EMAIL,
    headerName: COLUMNS.HEADER_NAME.EMAIL,
  },
  {
    field: COLUMNS.FIELDS.CLASS,
    headerName: COLUMNS.HEADER_NAME.CLASS,
  },
  {
    field: COLUMNS.FIELDS.GENDER,
    headerName: COLUMNS.HEADER_NAME.GENDER,
  },
];

export const COLUMNS_TEACHER = [
  {
    field: COLUMNS.FIELDS.NAME,
    headerName: COLUMNS.HEADER_NAME.NAME,
  },
  {
    field: COLUMNS.FIELDS.SUBJECT,
    headerName: COLUMNS.HEADER_NAME.SUBJECT,
  },
  {
    field: COLUMNS.FIELDS.CLASS,
    headerName: COLUMNS.HEADER_NAME.CLASS,
  },
  {
    field: COLUMNS.FIELDS.EMAIL,
    headerName: COLUMNS.HEADER_NAME.EMAIL,
  },
  {
    field: COLUMNS.FIELDS.GENDER,
    headerName: COLUMNS.HEADER_NAME.GENDER,
  },
];
