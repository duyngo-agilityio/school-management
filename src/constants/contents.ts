export const NOTFOUND = {
  STUDENT: {
    TITLE: 'No Students at this time',
    DESCRIPTION: 'Students will appear here after they enroll in your school.',
    BACK_BUTTON: 'Back to Student List'
  },
  TEACHER: {
    TITLE: 'No Teachers at this time',
    DESCRIPTION: 'Teachers will appear here after they enroll in your school.',
    BACK_BUTTON: 'Back to Teacher List'
  },
};
