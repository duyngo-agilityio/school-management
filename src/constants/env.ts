export const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL!;
export const IMG_KEY = process.env.NEXT_PUBLIC_IMG_KEY;
