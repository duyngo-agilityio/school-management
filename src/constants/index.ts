export * from './path';
export * from './placeholder';
export * from './colors';
export * from './columns';
export * from './messages';
export * from './options';
export * from './regex';
export * from './contents';
export * from './routes';
