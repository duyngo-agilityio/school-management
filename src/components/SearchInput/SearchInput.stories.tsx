import type { Meta, StoryObj } from '@storybook/react';

// Components
import SearchInput from '.';
import { INPUT_PLACEHOLDER } from '@/constants';
import { fn } from '@storybook/test';

const meta: Meta<typeof SearchInput> = {
  title: 'components/SearchInput',
  component: SearchInput,
  tags: ['autodocs'],
  argTypes: {
    placeholder: {
      description:
        'The placeholder text displayed inside the input field when it is empty',
    },
    defaultValue: {
      description: 'The initial value of the input field',
    },
    onSearch: {
      description:
        'A callback function triggered every time the input value changes',
    },
    disableInput: {
      description:
        'Disables the input field when set to true. The default value is false',
    },
  },
};

export default meta;
type Story = StoryObj<typeof SearchInput>;

export const Default: Story = {
  args: {
    placeholder: INPUT_PLACEHOLDER.STUDENT,
    defaultValue: '',
    onSearch: fn(),
    disableInput: false,
  },
};

export const Disabled: Story = {
  args: {
    placeholder: INPUT_PLACEHOLDER.STUDENT,
    defaultValue: 'Anna',
    onSearch: fn(),
    disableInput: true,
  },
};
