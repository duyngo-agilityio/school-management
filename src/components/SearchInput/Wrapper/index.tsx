'use client';

// Libs
import { usePathname, useRouter, useSearchParams } from 'next/navigation';
import { useCallback, useMemo } from 'react';
import { useDebouncedCallback } from 'use-debounce';

// Components
import SearchInput from '..';

interface SearchInputWrapperProps {
  placeholder: string;
  disableInput?: boolean;
}

const SearchInputWrapper = ({
  placeholder,
  disableInput,
}: SearchInputWrapperProps) => {
  const searchParams = useSearchParams();
  const { replace } = useRouter();
  const pathname = usePathname();

  const searchValue = useMemo(
    () => searchParams.get('q') || '',
    [searchParams],
  );

  const handleSearchValue = useCallback(
    useDebouncedCallback((value: string) => {
      const params = new URLSearchParams(searchParams);

      if (value) {
        params.set('q', value);
      } else {
        params.delete('q');
      }

      replace(`${pathname}?${params.toString()}`);
    }, 500),
    [pathname, replace, searchParams],
  );

  return (
    <SearchInput
      defaultValue={searchValue}
      placeholder={placeholder}
      onSearch={handleSearchValue}
      disableInput={disableInput}
    />
  );
};

export default SearchInputWrapper;
