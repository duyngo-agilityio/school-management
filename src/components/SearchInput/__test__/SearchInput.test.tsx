import { fireEvent, render } from '@testing-library/react';

// Components
import SearchInput from '..';

// Constants
import { INPUT_PLACEHOLDER } from '@/constants';

const mockOnSearch = jest.fn();
const mockOnClick = jest.fn();

const SearchInputProps = {
  placeholder: INPUT_PLACEHOLDER.TEACHER,
  defaultValue: '',
  onSearch: mockOnSearch,
  onClick: mockOnClick,
  disableInput: false,
};

const SearchInputComponent = () =>
  render(<SearchInput {...SearchInputProps} />);

describe('SearchInput component', () => {
  it('should match snapshot for SearchInput', () => {
    const { container } = SearchInputComponent();

    expect(container).toMatchSnapshot();
  });

  it('calls onSearch when input value changes', () => {
    const { getByPlaceholderText } = SearchInputComponent();

    const inputElement = getByPlaceholderText(INPUT_PLACEHOLDER.TEACHER);

    fireEvent.change(inputElement, { target: { value: 'math' } });

    expect(mockOnSearch).toHaveBeenCalledWith('math');
  });
});
