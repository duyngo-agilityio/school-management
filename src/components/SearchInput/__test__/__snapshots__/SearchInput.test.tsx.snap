// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component should match snapshot for SearchInput 1`] = `
<div>
  <div
    class="chakra-input__group css-ck76jc"
    data-group="true"
  >
    <div
      class="chakra-input__left-element css-o84xzr"
      style="cursor: pointer;"
    >
      <svg
        class="chakra-icon css-nyt2fo"
        focusable="false"
        viewBox="0 0 17 16"
      >
        <g
          clip-path="url(#clip0_16_102)"
        >
          <path
            d="M12.0725 10.344C13.0408 9.0227 13.4745 7.3845 13.2869 5.75716C13.0992 4.12982 12.304 2.63335 11.0603 1.56714C9.81666 0.50093 8.2163 -0.0563875 6.57938 0.00668566C4.94247 0.0697588 3.38973 0.748571 2.23182 1.90732C1.0739 3.06606 0.396199 4.61929 0.334297 6.25624C0.272396 7.8932 0.830858 9.49317 1.89796 10.7361C2.96505 11.9789 4.46209 12.7731 6.08957 12.9596C7.71704 13.1461 9.35493 12.7112 10.6755 11.742H10.6745C10.7045 11.782 10.7365 11.82 10.7725 11.857L14.6225 15.707C14.8101 15.8946 15.0644 16.0001 15.3297 16.0002C15.595 16.0003 15.8494 15.895 16.037 15.7075C16.2247 15.52 16.3302 15.2656 16.3303 15.0004C16.3303 14.7351 16.2251 14.4806 16.0375 14.293L12.1875 10.443C12.1518 10.4068 12.1134 10.3734 12.0725 10.343V10.344ZM12.3305 6.50001C12.3305 7.22228 12.1883 7.93748 11.9119 8.60477C11.6355 9.27206 11.2304 9.87837 10.7196 10.3891C10.2089 10.8998 9.6026 11.3049 8.93531 11.5813C8.26802 11.8577 7.55282 12 6.83055 12C6.10828 12 5.39308 11.8577 4.72579 11.5813C4.0585 11.3049 3.45218 10.8998 2.94146 10.3891C2.43074 9.87837 2.02561 9.27206 1.74921 8.60477C1.47281 7.93748 1.33055 7.22228 1.33055 6.50001C1.33055 5.04132 1.91001 3.64237 2.94146 2.61092C3.97291 1.57947 5.37186 1.00001 6.83055 1.00001C8.28924 1.00001 9.68819 1.57947 10.7196 2.61092C11.7511 3.64237 12.3305 5.04132 12.3305 6.50001Z"
            fill="currentColor"
          />
        </g>
        <defs>
          <clippath
            id="clip0_16_102"
          >
            <rect
              fill="white"
              height="16"
              transform="translate(0.330566)"
              width="16"
            />
          </clippath>
        </defs>
      </svg>
    </div>
    <input
      class="chakra-input css-nqkltn"
      placeholder="Search for a teacher by name or email"
      value=""
    />
    <div
      class="chakra-input__right-element css-dp9r5f"
    >
      <svg
        baseProfile="tiny"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        style="cursor: pointer;"
        version="1.2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 3c-4.963 0-9 4.038-9 9s4.037 9 9 9 9-4.038 9-9-4.037-9-9-9zm0 16c-3.859 0-7-3.14-7-7s3.141-7 7-7 7 3.14 7 7-3.141 7-7 7zM12.707 12l2.646-2.646c.194-.194.194-.512 0-.707-.195-.194-.513-.194-.707 0l-2.646 2.646-2.646-2.647c-.195-.194-.513-.194-.707 0-.195.195-.195.513 0 .707l2.646 2.647-2.646 2.646c-.195.195-.195.513 0 .707.097.098.225.147.353.147s.256-.049.354-.146l2.646-2.647 2.646 2.646c.098.098.226.147.354.147s.256-.049.354-.146c.194-.194.194-.512 0-.707l-2.647-2.647z"
        />
      </svg>
    </div>
  </div>
</div>
`;
